bindings/python/packaging/common_setup.py

```diff
@@ -0,0 +1,96 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import platform
+import setuptools
+import sys
+from datetime import date
+
+
+def get_package_dir():
+  if "CMAKE_BUILD_ROOT" in os.environ:
+    cmake_build_root = os.environ["CMAKE_BUILD_ROOT"]
+    print("Using CMake build root:", cmake_build_root)
+    pkg_dir = os.path.join(cmake_build_root, "bindings", "python")
+  else:
+    print("ERROR: No build directory specified. Set one of these variables:")
+    print("  CMAKE_BUILD_ROOT=/path/to/cmake/build")
+    sys.exit(1)
+
+  if not os.path.exists(pkg_dir):
+    print("ERROR: Package path does not exist:", pkg_dir)
+    sys.exit(1)
+  return pkg_dir
+
+
+def get_default_date_version():
+  today = date.today()
+  return today.strftime("%Y%m%d")
+
+
+def get_setup_defaults(sub_project, description):
+  return {
+    "name": "google-iree-%s-pkg" % (sub_project,),
+    "version": get_default_date_version(),
+    "author": "The IREE Team at Google",
+    "author_email": "iree-discuss@googlegroups.com",
+    "description": description,
+    "long_description": description,
+    "long_description_content_type": "text/plain",
+    "url": "https://github.com/google/iree",
+    "package_dir": {"": get_package_dir()},
+    "classifiers": [
+        "Programming Language :: Python :: 3",
+        "License :: OSI Approved :: Apache License",
+        "Operating System :: OS Independent",
+        "Development Status :: 3 - Alpha",
+    ],
+    "python_requires": ">=3.6",
+  }
+
+
+def get_native_file_extension():
+  if platform.system() == "Windows":
+    return "pyd"
+  elif platform.system() == "Darwin":
+    return "dylib"
+  else:
+    return "so"
+  
+
+def setup(**kwargs):
+  # See: https://stackoverflow.com/questions/45150304/how-to-force-a-python-wheel-to-be-platform-specific-when-building-it
+  try:
+      from wheel.bdist_wheel import bdist_wheel as _bdist_wheel
+      class bdist_wheel(_bdist_wheel):
+          def finalize_options(self):
+              _bdist_wheel.finalize_options(self)
+              self.root_is_pure = False
+  except ImportError:
+      bdist_wheel = None  
+
+  # Need to include platform specific extensions binaries:
+  #  Windows: .pyd
+  #  MacOS: .dylib
+  #  Other: .so
+  # Unfortunately, bazel is imprecise and scatters .so files around, so
```

Comment:
Kind of confused by the references to `CMAKE_BUILD_ROOT` and `bazel` in these files. Are both build systems involved? Are you thinking we'll use CMake for the non-TF packages, then Bazel for the TF package?

Reply:
In the end, all setuptools needs is a materialized output directory of artifacts, including source files and native extensions. On the CMake side, it is nice enough to produce a mirror of the source tree, so I elected to just have the user-provided part here be the build root. On the bazel-side, we will need a dummy py_binary target that builds all of the python packages we wish to deploy. Then we'll need to arrange to add its appropriately dereferenced runfiles to the python path (not unlike what the current colab launch script does). Since this will be a script that us (or bots) will be running quite a bit, I wanted to keep some big variable "knobs" here that do the right thing based on which build system is used (i.e. easy to just set a CMAKE_BUILD_ROOT var at the top of some instructions and then have copy-pastable build and package commands). We can always revisit.

And to answer your question: I don't want to put any more pressure on the bazel build for OSS artifacts than is necessary. That means only using it for the TF-connected portions. I think we can bridge the worlds with a couple of environment variables.